### PR TITLE
Medical kit now uses overheal on eligible attached limbs

### PR DESCRIPTION
**Pull request by Anasari**
If there's any remaining heal left after healing a targeted limb using a medkit, the remaining overheal would be applied to eligible limbs. This makes the advanced trauma & burn kit a bit stronger and more worth using. 

The eligible limbs are the limb's children (If any) and their parent. Starting from their children first.

EDIT:
For clarification, the advanced kit heals 25 (burn and trauma), and how this PR work is that if you have LEFTOVER HEAL (not use) from a single use, it'll be applied to eligible limbs. 

As an example. A kit with 5 uses and 25 brute, is applied on a right arm with 20 damage. There's a right hand with 20 damage.

Upon using it, ONE USE is deducted from the kit and the right arm would have 0 damage. The remaining bandage get applied to the right hand, and it now has 15 damage.

It doesn't force you to use up any additional uses - It just saves the "Spillover heal" so to speak.

tl;dr - Uses overheal on attached limbs if any

EDIT2:
Better worded CL thanks to Citinited

🆑 
tweak: Assuming there is heal left, bandages and ointments used on limbs now heal both the limb and the hand / foot. Bandages and ointments used on the torso now heal head, arms, and lower body. The one on lower body heals the legs too.
/🆑

### CODE REVIEW

**Comment by Ty-Omaha** [comment]
If people only have a few uses left, and they want to use it on specific parts like their head and hands they wouldn't be able to if a connecting part was damaged too.

There's little reason to use these over patches, cryotubes, or a sleeper in medbay. 

These are good on the go (so they'll be limited in supply), or for surgery.

**Comment by Citinited** [comment]
For reference, how much do these normally heal?

**Comment by Anasari** [comment]
Proof my communication skills are bad.

For clarification, the advanced kit heals 25 (burn and trauma), and how this PR work is that if you have LEFTOVER HEAL (not use) from a single use, it'll be applied to eligible limbs. 

As an example. A kit with 5 uses and 25 brute, is applied on a right arm with 20 damage. There's a right hand with 20 damage.

Upon using it, ONE USE is deducted from the kit and the right arm would have 0 damage. The remaining bandage get applied to the right hand, and it now has 15 damage.

It doesn't force you to use up any additional uses - It just saves the "Spillover heal" so to speak.

**Comment by Ty-Omaha** [comment]
ah

File: code/game/objects/items/stacks/medical.dm [hunk 1]
```diff
@@ -104,7 +104,27 @@
 			if(stop_bleeding)
 				if(!H.bleedsuppress) //so you can't stack bleed suppression
 					H.suppress_bloodloss(stop_bleeding)
+
+			var/rembrute = max(0, heal_brute - affecting.brute_dam)
+			var/remburn = max(0, heal_burn - affecting.burn_dam)
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: Do bruise packs even heal burn?

File: code/game/objects/items/stacks/medical.dm [hunk 2]
```diff
@@ -104,7 +104,27 @@
...
+			var/rembrute = max(0, heal_brute - affecting.brute_dam)
+			var/remburn = max(0, heal_burn - affecting.burn_dam)
+			var/nrembrute = rembrute
+			var/nremburn	= remburn
 			affecting.heal_damage(heal_brute, heal_burn)
+			if(affecting.body_part in list(ARM_LEFT, ARM_RIGHT, LEG_LEFT, LEG_RIGHT, UPPER_TORSO, LOWER_TORSO))
+				for(var/obj/item/organ/external/E in affecting.children)
+					if(rembrute <= 0 && remburn <= 0) // Make sure there's actually overheal left and the organ is damaged before bothering
+						break
+					else if(E.status && ORGAN_ROBOT || E.open == 1) // Ignore robotic or open limb
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: `E.open` will suffice, `E.open == 1` is redundant

File: code/game/objects/items/stacks/medical.dm [hunk 3]
```diff
@@ -104,7 +104,27 @@
...
+						continue
+					else if(E.brute_dam == 0 && E.burn_dam == 0) // Ignore undamaged limb
+						continue
+					nrembrute = rembrute - E.brute_dam // Deduct the healed damage from the remain
+					nremburn = remburn - E.burn_dam
+					E.heal_damage(rembrute, remburn)
+					rembrute = nrembrute
+					remburn = nremburn
+					user.visible_message("<span class='green'>[user] bandages the wounds on [H]'s [E.name] with the remaining bandages.</span>", \
+										 "<span class='green'>You bandages the wounds on [H]'s [E.name] with the remaining bandages.</span>" )
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: `You bandage` not `You bandages`

File: code/game/objects/items/stacks/medical.dm [hunk 4]
```diff
@@ -144,8 +164,28 @@
 			affecting.germ_level = 0
 
 			user.visible_message("<span class='green'>[user] salves the wounds on [H]'s [affecting.name].</span>", \
-							 	 "<span class='green'>You salve the wounds on [H]'s [affecting.name].</span>" )
+								 "<span class='green'>You salves the wounds on [H]'s [affecting.name].</span>" )
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: See above, `salve` not `salves`

File: code/game/objects/items/stacks/medical.dm [hunk 5]
```diff
@@ -144,8 +164,28 @@
 			affecting.germ_level = 0
 
 			user.visible_message("<span class='green'>[user] salves the wounds on [H]'s [affecting.name].</span>", \
-							 	 "<span class='green'>You salve the wounds on [H]'s [affecting.name].</span>" )
+								 "<span class='green'>You salves the wounds on [H]'s [affecting.name].</span>" )
+
+			var/rembrute = max(0, heal_brute - affecting.brute_dam)
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: See above, I'm confident ointment does jack shit for brute damage

File: code/game/objects/items/stacks/medical.dm [hunk 6]
```diff
@@ -104,7 +104,27 @@
 			if(stop_bleeding)
 				if(!H.bleedsuppress) //so you can't stack bleed suppression
 					H.suppress_bloodloss(stop_bleeding)
+
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: You should look into making both of these into a single proc so that you aren't duplicating similar code

**Comment by davidchan13** [comment]
Is change also going to affect Welders/Cable coils and Prosthetic/IPC limbs?

File: code/game/objects/items/stacks/medical.dm [hunk 7]
```diff
@@ -144,8 +164,28 @@
 			affecting.germ_level = 0
 
 			user.visible_message("<span class='green'>[user] salves the wounds on [H]'s [affecting.name].</span>", \
-							 	 "<span class='green'>You salve the wounds on [H]'s [affecting.name].</span>" )
+								 "<span class='green'>You salves the wounds on [H]'s [affecting.name].</span>" )
+
+			var/rembrute = max(0, heal_brute - affecting.brute_dam)
+			var/remburn = max(0, heal_burn - affecting.burn_dam)
+			var/nrembrute = rembrute
```
Comment (Citinited) [inline code/game/objects/items/stacks/medical.dm]: What do these variables (`nrembrute` and `nremburn`) actually do? It doesn't seem to be referenced usefully anywhere, all they seem to do is be equal to `rembrute` and `remburn`.

**Comment by FalseIncarnate** [comment]
@davidchan13 These changes do nothing to Cable/Welder repairs.

Those were somewhat updated (mostly the Cable repairing) in #5490 so that cable usage scales with damage healed, so you'd at worst be wasting just shy of 3 damage worth of healing (1 cable piece, which is far more common than these kits).

Welder healing would be a more likely source of healing waste (attempts to heal 15 damage per use) than the cables, and this is also probably fine as welder fuel is practically everywhere (healing with the welder consumes 1u of fuel), so it is not really a limited resource like the kits either.

**Comment by davidchan13** [comment]
@FalseIncarnate I bring it up not because of resource conservation (healing items aren't hard to come by, brute packs might be somewhat rare but better alternatives can be made readily in medical), but due to the time sync that it is to heal limbs when you've suffered whole body wounds, targeting each limb manually with bruise packs/ointment and prosthetics with welders/cable/nanopaste, excess healing automatically going to adjacent limbs would be a huge QOL improvement. As it stands, this PR is an indirect nerf to prosthetics and IPCs as now all their methods of healing are worse and more time consuming.

"Wire and Welder Fuel is common" doesn't fly with me as an excuse. MedChem, SciChem and Botany can push out more healing items in an hour than there is cable and welders on the station.

**Comment by FalseIncarnate** [comment]
I wasn't using it as an excuse, I was using it as reasoning for why they weren't given the same treatment here as the brute/burn trauma kits.

This is by no means an indirect nerf to anything, because this doesn't nerf anything or change any values. This is purely a buff to the trauma kits. IPCs and prosthetics are not meant to be a 1:1 with organics, and frankly we are not aiming to make them so.

If anything is going to get the treatment of the trauma kits for prosthetics, it should be nanopaste, as that is effectively a prosthetic trauma kit (it is even a stack like them).

I'd also avoid trying to bring chems into this discussion as that is an entirely separate system than what is being dealt with here. Chems will always be more convenient since they don't require specific limb targeting anyways, can heal over time, and are practically infinite supply for all but the rarest of chems. Bring them up just muddies the water of what is otherwise a straightforward PR discussion.

File: code/game/objects/items/stacks/medical.dm [hunk 8]
```diff
@@ -77,6 +78,41 @@
...
+						 "<span class='green'>You [healverb] the wounds on [H]'s [affecting.name].</span>" )
+
+	var/rembrute = max(0, heal_brute - affecting.brute_dam)
+	var/remburn = max(0, heal_burn - affecting.burn_dam)
+	var/nrembrute = rembrute
+	var/nremburn	= remburn
+	affecting.heal_damage(heal_brute, heal_burn)
+	var/list/achildlist = affecting.children.Copy()
+	var/parenthealed = 0
+	while(rembrute + remburn > 0): // Don't bother if there's not enough leftover heal
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: what's with the colons?

File: code/game/objects/items/stacks/medical.dm [hunk 9]
```diff
@@ -77,6 +78,41 @@
...
+/obj/item/stack/medical/proc/heal(mob/living/M, mob/user)
+	var/mob/living/carbon/human/H = M
+	var/obj/item/organ/external/affecting = H.get_organ(user.zone_sel.selecting)
+	user.visible_message("<span class='green'>[user] [healverb]s the wounds on [H]'s [affecting.name].</span>", \
+						 "<span class='green'>You [healverb] the wounds on [H]'s [affecting.name].</span>" )
+
+	var/rembrute = max(0, heal_brute - affecting.brute_dam)
+	var/remburn = max(0, heal_burn - affecting.burn_dam)
+	var/nrembrute = rembrute
+	var/nremburn	= remburn
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: a tab on one line, a space on the other?

File: code/game/objects/items/stacks/medical.dm [hunk 10]
```diff
@@ -77,6 +78,41 @@
...
+	affecting.heal_damage(heal_brute, heal_burn)
+	var/list/achildlist = affecting.children.Copy()
+	var/parenthealed = 0
+	while(rembrute + remburn > 0): // Don't bother if there's not enough leftover heal
+		var/obj/item/organ/external/E
+		if(achildlist.len):
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: colon

File: code/game/objects/items/stacks/medical.dm [hunk 11]
```diff
@@ -77,6 +78,41 @@
...
+			E = pick(achildlist) // Pick a random children
+			achildlist -= E
+		else if(affecting.parent && !parenthealed): // If there's a parent and no healing attempt was made on it
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: colon

File: code/game/objects/items/stacks/medical.dm [hunk 12]
```diff
@@ -77,6 +78,41 @@
...
+	user.visible_message("<span class='green'>[user] [healverb]s the wounds on [H]'s [affecting.name].</span>", \
+						 "<span class='green'>You [healverb] the wounds on [H]'s [affecting.name].</span>" )
+
+	var/rembrute = max(0, heal_brute - affecting.brute_dam)
+	var/remburn = max(0, heal_burn - affecting.burn_dam)
+	var/nrembrute = rembrute
+	var/nremburn	= remburn
+	affecting.heal_damage(heal_brute, heal_burn)
+	var/list/achildlist = affecting.children.Copy()
+	var/parenthealed = 0
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: `= FALSE`, since you use this as a boolean

File: code/game/objects/items/stacks/medical.dm [hunk 13]
```diff
@@ -77,6 +78,41 @@
...
+	while(rembrute + remburn > 0): // Don't bother if there's not enough leftover heal
+		var/obj/item/organ/external/E
+		if(achildlist.len):
+			E = pick(achildlist) // Pick a random children
+			achildlist -= E
+		else if(affecting.parent && !parenthealed): // If there's a parent and no healing attempt was made on it
+			E = affecting.parent
+			parenthealed = 1
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: `= TRUE`

File: code/game/objects/items/stacks/medical.dm [hunk 14]
```diff
@@ -77,6 +78,41 @@
...
+	var/remburn = max(0, heal_burn - affecting.burn_dam)
+	var/nrembrute = rembrute
+	var/nremburn	= remburn
+	affecting.heal_damage(heal_brute, heal_burn)
+	var/list/achildlist = affecting.children.Copy()
+	var/parenthealed = 0
+	while(rembrute + remburn > 0): // Don't bother if there's not enough leftover heal
+		var/obj/item/organ/external/E
+		if(achildlist.len):
+			E = pick(achildlist) // Pick a random children
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: `E = pick_n_take(achildlist)`

File: code/game/objects/items/stacks/medical.dm [hunk 15]
```diff
@@ -77,6 +78,41 @@
...
+		else if(affecting.parent && !parenthealed): // If there's a parent and no healing attempt was made on it
+			E = affecting.parent
+			parenthealed = 1
+		else:
+			break // If the organ have no child left and no parent / parent healed, break
+		if(E.status & ORGAN_ROBOT || E.open) // Ignore robotic or open limb
+			continue
+		else if(!E.brute_dam && !E.burn_dam) // Ignore undamaged limb
+			continue
+		nrembrute = rembrute - E.brute_dam // Deduct the healed damage from the remain
```
Comment (Crazylemon64) [inline code/game/objects/items/stacks/medical.dm]: Should these values be `max`ed with 0

Comment (Anasari) [reply code/game/objects/items/stacks/medical.dm]: They're maxed to zero since heal_damage doesn't have a safety check for negative values. You can pass in a negative value and have it deal damage. 

	brute_dam = max(brute_dam - brute, 0)
	burn_dam  = max(burn_dam - burn, 0)

Comment (Crazylemon64) [reply code/game/objects/items/stacks/medical.dm]: So if you had a lower body with 5 points of brute damage, and a leg with 30 points of burn damage, and another leg with 10 points of brute
and then you used a brute kit with 25 points
You'd heal the lower body, `25-5 = 20` points of overflow
You check the leg with 30 burn
`nrembrute = 20 - 0`
`nremburn = 0 - 30`
Then you hit the `while` check again, `20 + (-30) > 0 == FALSE`
So the loop terminates, and you don't get to hit the leg with 10 brute

**Comment by Crazylemon64** [comment]
dangit travis
